File: web/src/types/matchup.ts
export interface OpponentPokemon {
  name: string;
}

export type MatchFormat = "ladder" | "bo1" | "bo3" | "tournament" | "friendly";
export type CloseType = "blowout" | "close" | "comeback" | "standard";

export interface Matchup {
  id: string;
  user_id: string;
  my_team_id: string | null;
  opponent_team_data: OpponentPokemon[] | null;
  lead_pair: string[] | null;
  outcome: "win" | "loss";
  notes: string | null;
  played_at: string;
  format?: MatchFormat | null;
  tags?: string[];
  close_type?: CloseType | null;
  mvp_pokemon?: string | null;
  my_team_actual?: string[] | null;
  replay_url?: string | null;
  opponent_name?: string | null;
  opponent_rating?: number | null;
  event_name?: string | null;
  round_label?: string | null;
  game_number?: number | null;
  set_id?: string | null;
  opponent_lead_pair?: string[] | null;
  opponent_selected_four?: string[] | null;
  my_selected_four?: string[] | null;
  loss_reason?: string | null;
  win_condition?: string | null;
  key_turn?: string | null;
  adjustment_note?: string | null;
}

export interface MatchupCreate {
  my_team_id?: string | null;
  opponent_team_data: OpponentPokemon[];
  lead_pair?: string[];
  outcome: "win" | "loss";
  notes?: string;
  format?: MatchFormat;
  tags?: string[];
  close_type?: CloseType;
  mvp_pokemon?: string;
  my_team_actual?: string[];
  replay_url?: string;
  opponent_name?: string;
  opponent_rating?: number;
  event_name?: string;
  round_label?: string;
  game_number?: number;
  set_id?: string;
  opponent_lead_pair?: string[];
  opponent_selected_four?: string[];
  my_selected_four?: string[];
  loss_reason?: string;
  win_condition?: string;
  key_turn?: string;
  adjustment_note?: string;
}

export interface MatchupUpdate {
  opponent_team_data?: OpponentPokemon[];
  lead_pair?: string[];
  outcome?: "win" | "loss";
  notes?: string;
  format?: MatchFormat;
  tags?: string[];
  close_type?: CloseType;
  mvp_pokemon?: string;
  my_team_actual?: string[];
  replay_url?: string;
  opponent_name?: string;
  opponent_rating?: number;
  event_name?: string;
  round_label?: string;
  game_number?: number;
  set_id?: string;
  opponent_lead_pair?: string[];
  opponent_selected_four?: string[];
  my_selected_four?: string[];
  loss_reason?: string;
  win_condition?: string;
  key_turn?: string;
  adjustment_note?: string;
}

export interface MatchupListResponse {
  data: Matchup[];
  count: number;
}

export interface WinRateStat {
  label: string;
  wins: number;
  losses: number;
  total: number;
  win_rate: number;
}

export interface MatchupStats {
  overall: WinRateStat;
  by_team: WinRateStat[];
  by_opponent_pokemon: WinRateStat[];
  by_format?: WinRateStat[];
  by_tag?: WinRateStat[];
}

export interface FrequencyStat {
  label: string;
  count: number;
}

export interface PrepAction {
  label: string;
  detail: string;
  action: "benchmark_team" | "review_team" | "review_losses" | "log_match" | string;
}

export interface MatchupInsights {
  total_matches: number;
  recent: WinRateStat;
  worst_opponents: WinRateStat[];
  underperforming_teams: WinRateStat[];
  common_loss_reasons: FrequencyStat[];
  common_loss_tags: FrequencyStat[];
  prep_actions: PrepAction[];
}

File: api/app/models/matchup.py
from datetime import datetime

from pydantic import BaseModel, Field, model_validator

# Enum-like validators via regex (pydantic v2 style). Migration
# 20260419100000 enforces the same set via CHECK constraints in Postgres.
FORMAT_PATTERN = r"^(ladder|bo1|bo3|tournament|friendly)$"
CLOSE_TYPE_PATTERN = r"^(blowout|close|comeback|standard)$"


class OpponentPokemon(BaseModel):
    name: str
    """Title Case Pokemon name."""


class MatchupCreate(BaseModel):
    my_team_id: str | None = Field(None, description="UUID of saved team used")
    opponent_team_data: list[OpponentPokemon] = Field(
        min_length=1,
        max_length=6,
        description="Opponent's team (1-6 Pokemon)",
    )
    lead_pair: list[str] | None = Field(
        None,
        min_length=2,
        max_length=2,
        description="My lead pair Pokemon names",
    )
    outcome: str = Field(pattern=r"^(win|loss)$")
    notes: str | None = None
    format: str | None = Field(
        None,
        pattern=FORMAT_PATTERN,
        description="Match context: ladder | bo1 | bo3 | tournament | friendly",
    )
    tags: list[str] = Field(
        default_factory=list,
        description="Archetype/strategy tags (e.g. 'rain', 'trick-room')",
    )
    close_type: str | None = Field(
        None,
        pattern=CLOSE_TYPE_PATTERN,
        description="Post-match reflection: blowout | close | comeback | standard",
    )
    mvp_pokemon: str | None = Field(
        None, description="The Pokemon that carried (or failed) the match"
    )
    my_team_actual: list[str] | None = Field(
        None,
        max_length=6,
        description=(
            "Optional override of the actual lineup run in this match. "
            "NULL = same as the saved team's roster."
        ),
    )
    replay_url: str | None = None
    opponent_name: str | None = None
    opponent_rating: int | None = None
    event_name: str | None = None
    round_label: str | None = None
    game_number: int | None = None
    set_id: str | None = None
    opponent_lead_pair: list[str] | None = Field(None, max_length=2)
    opponent_selected_four: list[str] | None = Field(None, max_length=4)
    my_selected_four: list[str] | None = Field(None, max_length=4)
    loss_reason: str | None = None
    win_condition: str | None = None
    key_turn: str | None = None
    adjustment_note: str | None = None

    @model_validator(mode="after")
    def validate_team_source(self) -> "MatchupCreate":
        if not self.my_team_id and not self.my_team_actual:
            raise ValueError("Provide either my_team_id or my_team_actual")
        return self


class MatchupUpdate(BaseModel):
    opponent_team_data: list[OpponentPokemon] | None = None
    lead_pair: list[str] | None = None
    outcome: str | None = Field(None, pattern=r"^(win|loss)$")
    notes: str | None = None
    format: str | None = Field(None, pattern=FORMAT_PATTERN)
    tags: list[str] | None = None
    close_type: str | None = Field(None, pattern=CLOSE_TYPE_PATTERN)
    mvp_pokemon: str | None = None
    my_team_actual: list[str] | None = Field(None, max_length=6)
    replay_url: str | None = None
    opponent_name: str | None = None
    opponent_rating: int | None = None
    event_name: str | None = None
    round_label: str | None = None
    game_number: int | None = None
    set_id: str | None = None
    opponent_lead_pair: list[str] | None = Field(None, max_length=2)
    opponent_selected_four: list[str] | None = Field(None, max_length=4)
    my_selected_four: list[str] | None = Field(None, max_length=4)
    loss_reason: str | None = None
    win_condition: str | None = None
    key_turn: str | None = None
    adjustment_note: str | None = None


class MatchupResponse(BaseModel):
    id: str
    user_id: str
    my_team_id: str | None
    opponent_team_data: list[OpponentPokemon] | None
    lead_pair: list[str] | None
    outcome: str
    notes: str | None
    played_at: datetime
    format: str | None = None
    tags: list[str] = Field(default_factory=list)
    close_type: str | None = None
    mvp_pokemon: str | None = None
    my_team_actual: list[str] | None = None
    replay_url: str | None = None
    opponent_name: str | None = None
    opponent_rating: int | None = None
    event_name: str | None = None
    round_label: str | None = None
    game_number: int | None = None
    set_id: str | None = None
    opponent_lead_pair: list[str] | None = None
    opponent_selected_four: list[str] | None = None
    my_selected_four: list[str] | None = None
    loss_reason: str | None = None
    win_condition: str | None = None
    key_turn: str | None = None
    adjustment_note: str | None = None


class MatchupList(BaseModel):
    data: list[MatchupResponse]
    count: int


class WinRateStat(BaseModel):
    label: str
    wins: int
    losses: int
    total: int
    win_rate: float = Field(description="Win rate as percentage 0-100")


class MatchupStats(BaseModel):
    overall: WinRateStat
    by_team: list[WinRateStat]
    by_opponent_pokemon: list[WinRateStat]
    by_format: list[WinRateStat] = Field(default_factory=list)
    by_tag: list[WinRateStat] = Field(default_factory=list)
